// midori/lib/test/createMockRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bl = _interopRequireDefault(require("bl"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const createMockRequest = options => {
  let reqBodyStream;

  const ensureReqBody = () => {
    if (!reqBodyStream) {
      if (typeof options.body === 'string' || Buffer.isBuffer(options.body)) {
        reqBodyStream = (0, _bl.default)(options.body);
      } else if (options.body && typeof options.body.read === 'function') {
        // flowlint-next-line unclear-type: off
        reqBodyStream = options.body;
      } else if (typeof options.body === 'undefined') {
        reqBodyStream = (0, _bl.default)('');
      } else {
        throw new TypeError('Invalid request body given.');
      }
    }
  };

  const rawHeaders = options.headers || {};
  Object.keys(rawHeaders).forEach(k => {
    rawHeaders[k.toLowerCase()] = rawHeaders[k];
  });
  const req = {
    method: typeof options.method === 'string' ? options.method : 'GET',
    url: typeof options.url === 'string' ? options.url : '/',
    headers: rawHeaders,
    // flowlint-next-line unclear-type: off
    read: (...args) => {
      ensureReqBody();
      return reqBodyStream.read(...args);
    },
    // flowlint-next-line unclear-type: off
    pipe: (...args) => {
      ensureReqBody();
      return reqBodyStream.pipe(...args);
    },
    // flowlint-next-line unclear-type: off
    on: (...args) => {
      ensureReqBody();
      reqBodyStream.on(...args);
      return req;
    },
    // flowlint-next-line unclear-type: off
    once: (...args) => {
      ensureReqBody();
      reqBodyStream.on(...args);
      return req;
    },
    // flowlint-next-line unclear-type: off
    removeListener: (...args) => {
      ensureReqBody();
      reqBodyStream.removeListener(...args);
      return req;
    },
    connection: {
      encrypted: options.encrypted
    },
    // TODO: Consider stubbing more of this?
    socket: {
      _handle: options.offline === true ? null : 1
    },

    // flowlint-next-line unsafe-getters-setters:off
    get body() {
      ensureReqBody();
      return reqBodyStream;
    }

  };
  return req;
};

var _default = createMockRequest;
exports.default = _default;
