"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _send = _interopRequireDefault(require("./send"));

var _url = require("url");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const normalize2 = url => {
  if (typeof url === 'string') {
    return url;
  } else if (typeof url === 'object') {
    return (0, _url.format)(url);
  }

  throw new TypeError(`Invalid url: "${url}"`);
};

const normalize = (statusCode, url) => {
  return {
    statusCode,
    url: normalize2(url)
  };
};
/**
 * Redirect somewhere. Ends the app chain.
 * @param {?Number} statusCode The status code to send back, defaults to 302.
 * @param {String|Object|Function} url The url to redirect to.
 * @returns {App} App instance.
 */


const redirect = (...args) => {
  // TODO: FIXME: How to get flow happy with this?
  const {
    url,
    statusCode
  } = normalize( // $ExpectError
  args.length > 1 ? args[0] : 302, // $ExpectError
  args.length <= 1 ? args[0] : args[1]);
  return (0, _send.default)(statusCode, {
    Location: url
  }, '');
};

var _default = redirect;
exports.default = _default;
//# sourceMappingURL=redirect.js.map